File: Js_Solutions/stack/stack.js
// Create a stack data structure.The stack
// should be a class that contains 'push' and 'pop'
// methods. 

class Node {
    constructor(val) {
        this.value = val;
        this.next = null;
    }
}

class Stack {
    constructor() {
        this.first = null;
        this.last = null;
        this.size = 0;
    }

    push(val) {
        const value = new Node(val);
        if (this.size === 0) {
            this.first = value;
            this.last = value;
        } else {
            const temp = this.first;
            this.first = value;
            this.first.next = temp;
        }
        this.size++;
    }

    pop() {
        if (this.size === 0) {
            return null;
        }
        const temp = this.first;
        if (this.size === 1) {
            this.first = null;
            this.last = null;
        } else {
            this.first = this.first.next;
        }
        this.size--;
        return temp.value;
    }
}

const s = new Stack();
s.push('a');
console.log(s);
s.push('b');
console.log(s);
s.push('c');
console.log(s);
console.log(s.pop());
console.log(s); 

File: Python_Solutions/stack/stack.py
# Create a stack data structure.The stack
# should be a class that contains 'push' and 'pop'
# methods.

class Node:
    def __init__(self, val):
        self.value = val
        self.next = None


class Stack:
    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0

    def push(self, val):
        value = Node(val)

        if self.size == 0:
            self.first = value
            self.last = value
        else:
            temp = self.first
            self.first = value
            self.first.next = temp
        self.size += 1

    def pop(self):
        if self.size == 0:
            return None

        temp = self.first

        if self.size == 1:
            self.first = None
            self.last = None
        else:
            self.first = self.first.next
        self.size -= 1
        return temp.value


s = Stack()
s.push('a')
print(s)
s.push('b')
s.push('c')
print(s.pop())
print(s.pop())
print(s.pop())
print(s.pop())
